File: p2p-ws-sdk-dart/example/core_ws_rpc_discover_echo.dart
import "dart:io";

import "package:p2p_ws_sdk/src/core_compat/core_ws_client.dart";
import "package:p2p_ws_sdk/src/core_compat/ordinals.dart";
import "package:p2p_ws_sdk/src/core_compat/rpc_client.dart";

Future<void> main(List<String> args) async {
  final wsUrl = args.isNotEmpty ? args[0] : "ws://127.0.0.1:18089/p2p";
  final magic = args.length >= 2 ? int.parse(args[1]) : -252702961;
  final privateKeyPemPath = args.length >= 3 ? args[2] : null;
  final userId = args.length >= 4 ? args[3] : "example-user-id";
  final msg = args.length >= 5 ? args[4] : "hello from dart core_compat";
  final ordinalsPath = args.length >= 6 ? args[5] : "../p2p-ws-protocol/generated/p2p_command_ordinals.json";

  if (privateKeyPemPath == null) {
    stderr.writeln("need client private key pem path arg2");
    exitCode = 2;
    return;
  }

  final ordinals = loadCommandOrdinalsFromJsonFile(ordinalsPath);
  final pem = File(privateKeyPemPath).readAsStringSync();

  final c = CoreWsClient(CoreWsClientConfig(wsUrl: wsUrl, magic: magic), ordinals);
  await c.connect();
  try {
    await c.handshakeAndLogin(userId: userId, clientPrivateKeyPem: pem);
    final rpc = CoreRpcClient(c);

    final services = await rpc.discover(includeMethods: true);
    stdout.writeln("discover.services=${services.length}");
    for (final s in services.take(10)) {
      stdout.writeln("  - ${s.service} version=${s.version} methods=${s.methods.length}");
    }

    final health = await rpc.health(service: "p2p.rpc.echo.v1.EchoService");
    stdout.writeln("health.healthy=${health.healthy} ready=${health.ready} message=${health.message}");

    final echo = await rpc.echo(msg);
    stdout.writeln("echo.message=${echo.message} server_time=${echo.serverTime}");
  } finally {
    await c.close();
  }
}

File: p2p-ws-sdk-ts/demo/core_ws_rpc_discover_echo.ts
import path from "node:path"

import { CoreWsClient } from "../src/core_compat/CoreWsClient.js"

async function main(): Promise<void> {
  const wsUrl = process.argv[2] ?? "ws://127.0.0.1:18089/p2p"
  const magic = process.argv[3] ? Number(process.argv[3]) : -252702961
  const privateKeyPemPath = process.argv[4]
  const userId = process.argv[5] ?? "example-user-id"
  const msg = process.argv[6] ?? "hello from ts core_compat"

  if (!privateKeyPemPath) throw new Error("need client private key pem path argv4")

  const absKey = path.resolve(process.cwd(), privateKeyPemPath)
  const c = new CoreWsClient({ wsUrl, magic })
  await c.connect()
  try {
    await c.handshakeAndLogin(userId, absKey)
    const disc = await c.rpcDiscover("", true)
    const services = Array.isArray(disc.services) ? disc.services : []
    console.log("discover.services=", services.length)
    for (const svc of services.slice(0, 10)) {
      const methods = Array.isArray(svc.methods) ? svc.methods : []
      console.log("  -", svc.service, "version=", svc.version, "methods=", methods.length)
    }

    const health = await c.rpcHealth("p2p.rpc.echo.v1.EchoService")
    console.log("health.healthy=", health.healthy, "ready=", health.ready, "message=", health.message)

    const echo = await c.rpcEcho(msg)
    console.log("echo.message=", echo.message, "server_time=", echo.server_time)
  } finally {
    c.close()
  }
}

main().catch((e) => {
  console.error(e)
  process.exit(1)
})

